In [1]:
import numpy as np

## Question 1

In [2]:
np.__version__

'1.23.3'

## Question 2

In [3]:
import pandas as pd

In [4]:
df = pd.read_csv('data.csv', header = 0)

In [6]:
len(df)

11914

How many records are in the dataset?
Answer is 11914

## Question 3

In [7]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [8]:
count_maker = df.groupby(['Make'])['Make'].count()

In [9]:
count_maker.sort_values(ascending = False).head(3)

Make
Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

Who are the most frequent car manufacturers (top-3) according to the dataset?

Chevrolet, Ford, Volkswagen

## Question 4

In [10]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [21]:
audi_df = df[(df['Make'] == 'Audi')]

In [22]:
print(audi_df['Model'].unique())
print(audi_df['Model'].nunique())

['100' '200' '80' '90' 'A3' 'A4 allroad' 'A4' 'A5' 'A6' 'A7' 'A8'
 'allroad quattro' 'allroad' 'Cabriolet' 'Coupe' 'Q3' 'Q5' 'Q7' 'R8'
 'RS 4' 'RS 5' 'RS 6' 'RS 7' 'S3' 'S4' 'S5' 'S6' 'S7' 'S8' 'SQ5' 'TT RS'
 'TT' 'TTS' 'V8']
34


What's the number of unique Audi car models in the dataset?

34

## Question 5

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11914 entries, 0 to 11913
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Make               11914 non-null  object 
 1   Model              11914 non-null  object 
 2   Year               11914 non-null  int64  
 3   Engine Fuel Type   11911 non-null  object 
 4   Engine HP          11845 non-null  float64
 5   Engine Cylinders   11884 non-null  float64
 6   Transmission Type  11914 non-null  object 
 7   Driven_Wheels      11914 non-null  object 
 8   Number of Doors    11908 non-null  float64
 9   Market Category    8172 non-null   object 
 10  Vehicle Size       11914 non-null  object 
 11  Vehicle Style      11914 non-null  object 
 12  highway MPG        11914 non-null  int64  
 13  city mpg           11914 non-null  int64  
 14  Popularity         11914 non-null  int64  
 15  MSRP               11914 non-null  int64  
dtypes: float64(3), int64(5

In [39]:
df.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders        0
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

How many columns in the dataset have missing values?

5

## Question 6

In [36]:
df['Engine Cylinders'].median()

6.0

In [37]:
df['Engine Cylinders'].mode()

0    4.0
Name: Engine Cylinders, dtype: float64

In [38]:
df['Engine Cylinders'].fillna(df['Engine Cylinders'].mode()[0],inplace=True)

In [40]:
df['Engine Cylinders'].median()

6.0

Has it changed?

No

## Question 7

In [88]:
df.Make.unique()

array(['BMW', 'Audi', 'FIAT', 'Mercedes-Benz', 'Chrysler', 'Nissan',
       'Volvo', 'Mazda', 'Mitsubishi', 'Ferrari', 'Alfa Romeo', 'Toyota',
       'McLaren', 'Maybach', 'Pontiac', 'Porsche', 'Saab', 'GMC',
       'Hyundai', 'Plymouth', 'Honda', 'Oldsmobile', 'Suzuki', 'Ford',
       'Cadillac', 'Kia', 'Bentley', 'Chevrolet', 'Dodge', 'Lamborghini',
       'Lincoln', 'Subaru', 'Volkswagen', 'Spyker', 'Buick', 'Acura',
       'Rolls-Royce', 'Maserati', 'Lexus', 'Aston Martin', 'Land Rover',
       'Lotus', 'Infiniti', 'Scion', 'Genesis', 'HUMMER', 'Tesla',
       'Bugatti'], dtype=object)

In [89]:
lotus_df = df[(df['Make'] == 'Lotus')]

In [90]:
lotus_df.head()

Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   
3915  Lotus  Elise  2010  premium unleaded (required)      189.0   
3916  Lotus  Elise  2010  premium unleaded (required)      218.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact   Convertible           27   
3913  Exotic,High-Performance      Compact   Convertible           26   
3914  Exotic,High-Performance      Compact   Convertible           27   
3915  Exotic,High-Performance      Compact   Convertible           27   
3916  Exotic,High-Performance      Compact   Convertible           26   

      city mpg  Popularity   MSRP  
3912        21         613  43995  
3913        20         613  54990  
3914        21         613  47250  
3915        21         613  47250  
3916        20         613  54990

In [91]:
lotus_df = lotus_df[['Engine HP','Engine Cylinders']]

In [94]:
lotus_df = lotus_df.drop_duplicates()
lotus_df

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [102]:
x = lotus_df.to_numpy()
print(x)
print(x.T)

[[189.   4.]
 [218.   4.]
 [217.   4.]
 [350.   8.]
 [400.   6.]
 [276.   6.]
 [345.   6.]
 [257.   4.]
 [240.   4.]]
[[189. 218. 217. 350. 400. 276. 345. 257. 240.]
 [  4.   4.   4.   8.   6.   6.   6.   4.   4.]]


In [105]:
xTx = np.dot(x.T,x)
print(xTx)

[[7.31684e+05 1.34100e+04]
 [1.34100e+04 2.52000e+02]]


In [106]:
xTx_inv = np.linalg.inv(xTx)
print(xTx_inv)

[[ 5.53084235e-05 -2.94319825e-03]
 [-2.94319825e-03  1.60588447e-01]]


In [100]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [115]:
w = np.dot(np.dot(xTx_inv, x.T), y)
w

array([  4.59494481, -63.56432501])